Add decryption/encryption dedicated APIs

youennf opened the pull request:
Fixes https://github.com/w3c/webrtc-encoded-transform/issues/259


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/youennf/webrtc-insertable-streams/pull/288.html" title="Last updated on Dec 19, 2025, 11:17 AM UTC (5905a63)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webrtc-encoded-transform/288/47531dd...youennf:5905a63.html" title="Last updated on Dec 19, 2025, 11:17 AM UTC (5905a63)">Diff</a>

youennf committed:
Add decryption/encryption dedicated APIs

youennf committed:
Introduce sender and receiver transforms

guidou commented on index.bs:
```diff
@@ -81,15 +81,16 @@ It uses an additional API on {{RTCRtpSender}} and {{RTCRtpReceiver}} to
 <pre class="idl">
-typedef (SFrameTransform or RTCRtpScriptTransform) RTCRtpTransform;
+typedef (SFrameSenderTransform or RTCRtpScriptTransform) RTCRtpSenderTransform;
+typedef (SFrameReceiverTransform or RTCRtpScriptTransform) RTCRtpReceierTransform;
```
typo RTCRtpReceierTransform -> RTCRtpReceiverTransform

youennf replied:
Thanks, fixed

jan-ivar commented on index.bs:
```diff
@@ -81,15 +81,16 @@ It uses an additional API on {{RTCRtpSender}} and {{RTCRtpReceiver}} to
 <pre class="idl">
-typedef (SFrameTransform or RTCRtpScriptTransform) RTCRtpTransform;
+typedef (SFrameSenderTransform or RTCRtpScriptTransform) RTCRtpSenderTransform;
+typedef (SFrameReceiverTransform or RTCRtpScriptTransform) RTCRtpReceierTransform;
```
```suggestion
typedef (SFrameReceiverTransform or RTCRtpScriptTransform) RTCRtpReceiverTransform;
```

youennf replied:
```suggestion
typedef (SFrameReceiverTransform or RTCRtpScriptTransform) RTCRtpReceiverTransform;
```

jan-ivar commented on index.bs:
```diff
@@ -335,14 +354,40 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+1. If |keyID| is a {{bigint}} which cannot be represented as a integer between 0 and 2<sup>64</sup>-1 inclusive, [=reject=] |promise| with a {{RangeError}} exception and abort these steps.
+1.  Set |this|.`[[currentKeyId]]` to |keyId|.
+1. [=In parallel=], run the following steps:
+    1. Set |key| and |keyID| as key material to use for the SFrame transform encryption algorithm, as defined by [[RFC9605]].
```
Nit: "Set as" is a bit confusing to me. It's the reverse of (yet similar looking to) "set a to b" which usually means a = b, not b = a... How about:

```suggestion
    1. Set the SFrame transform encryption algorithm's key material to |key| and |keyID|, as defined by [[RFC9605]].
```

youennf commented on index.bs:
```diff
@@ -335,14 +354,40 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+    1. [=Resolve=] |promise| with undefined.
+1. Return |promise|.
+
+// FIXME: Should SFrameTransform receiver be made aware of the current key in use, so that it would call removeEncryptionKey appropriately.
```
```suggestion
```

guidou commented on index.bs:
```diff
@@ -335,14 +354,40 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+    1. [=Resolve=] |promise| with undefined.
+1. Return |promise|.
+
+// FIXME: Should SFrameTransform receiver be made aware of the current key in use, so that it would call removeEncryptionKey appropriately.
```
remove fixme and file an issue instead

youennf replied:
done

youennf commented on index.bs:
```diff
@@ -335,14 +354,40 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+1. Return |promise|.
+
+// FIXME: Should SFrameTransform receiver be made aware of the current key in use, so that it would call removeEncryptionKey appropriately.
+// Or should we add an option to let the UA remove the key automatically on new KeyID?
```
```suggestion
```

jan-ivar commented on index.bs:
```diff
@@ -335,14 +354,40 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+    1. [=Queue a task=] to [=resolve=] |promise| with undefined.
+1. Return |promise|.
+
+The <dfn method for="SFrameDecrypterKeyManager">addEncryptionKey(|key|, |keyID|)</dfn> method steps are:
```
```suggestion
The <dfn method for="SFrameDecrypterKeyManager">addDecryptionKey(|key|, |keyID|)</dfn> method steps are:
```

youennf committed:
Apply suggestions from code review

Removing fixme in favour of https://github.com/w3c/webrtc-encoded-transform/issues/289

youennf committed:
Update index.bs

jan-ivar commented on index.bs:
```diff
@@ -335,14 +354,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+    1. [=Resolve=] |promise| with undefined.
+1. Return |promise|.
+
+The <dfn method for="SFrameDecrypterKeyManager">removeEncryptionKey(|key|, |keyID|)</dfn> method steps are:
```
```suggestion
The <dfn method for="SFrameDecrypterKeyManager">removeDecryptionKey(|key|, |keyID|)</dfn> method steps are:
```

jan-ivar commented on index.bs:
```diff
@@ -335,14 +354,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+1. If |keyID| is a {{bigint}} which cannot be represented as a integer between 0 and 2<sup>64</sup>-1 inclusive, [=reject=] |promise| with a {{RangeError}} exception, and abort these steps..
+1. [=In parallel=], run the following steps:
+    1. Let |keyStore| be the key store used for the SFrame transform algorithm, as defined by [[RFC9605]].
+    1. Set an entry to |keyStore| with |keyId| as key and |keyValue| as value. This overrides any existing entry to |keyId|.
```
I see "key store" matches language in RFC9605, which is nice.

But I also wonder if we could use infra here. Is it an https://infra.spec.whatwg.org/#ordered-map ?

youennf replied:
Done

youennf committed:
Update index.bs

Co-authored-by: Jan-Ivar Bruaroey <jan-ivar@users.noreply.github.com>

youennf committed:
Update index.bs

Co-authored-by: Jan-Ivar Bruaroey <jan-ivar@users.noreply.github.com>

youennf committed:
Update index.bs

Co-authored-by: Jan-Ivar Bruaroey <jan-ivar@users.noreply.github.com>

youennf committed:
Reuse infra map set

youennf committed:
Fix bikeshed link

youennf committed:
Fix video frame type

jan-ivar commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
-    2. If setting the key material fails, [=reject=] |promise| with an {{InvalidModificationError}} exception and abort these steps.
-    3. [=Resolve=] |promise| with undefined.
-4. Return |promise|.
+1. If |keyId| is <code>undefined</code>, run the following steps:
```
Note to self to find webIDL's preferred prose for optional method args

jan-ivar replied:
Maybe make `keyId` the method argument mandatory instead?

youennf replied:
done

jan-ivar commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+    1. Let |keyStore| be the key store used for the SFrame transform algorithm, as defined by [[RFC9605]].
+    1. [=map/set|Set=] |keyStore|[|keyId|] to |keyValue|.
+    1. If setting the key material fails, [=queue a task=] to [=reject=] |promise| with an {{InvalidModificationError}} exception and abort these steps.
+    1. [=Resolve=] |promise| with undefined.
```
Rewrite to queue a task to resolve promise as well

youennf replied:
done

youennf commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+    1. Let |keyStore| be the key store used for the SFrame transform algorithm, as defined by [[RFC9605]].
+    1. [=map/set|Set=] |keyStore|[|keyId|] to |keyValue|.
+    1. If setting the key material fails, [=queue a task=] to [=reject=] |promise| with an {{InvalidModificationError}} exception and abort these steps.
+    1. [=Resolve=] |promise| with undefined.
```
```suggestion
    1. [=Queue a task=] to [=resolve=] |promise| with undefined.
```

jan-ivar commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+1. [=In parallel=], run the following steps:
+    1. Let |keyStore| be the key store used for the SFrame transform algorithm, as defined by [[RFC9605]].
+    1. Remove the entry of |keyStore| at |keyId| if it exits.
+    1. [=Resolve=] |promise| with undefined.
```
ditto

youennf replied:
done

youennf committed:
Update index.bs

youennf commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+1. [=In parallel=], run the following steps:
+    1. Let |keyStore| be the key store used for the SFrame transform algorithm, as defined by [[RFC9605]].
+    1. Remove the entry of |keyStore| at |keyId| if it exits.
+    1. [=Resolve=] |promise| with undefined.
```
```suggestion
    1. [=Queue a task=] to [=resolve=] |promise| with undefined.
```

youennf committed:
Update index.bs

youennf commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+
+The <dfn method for="SFrameDecrypterKeyManager">addDecryptionKey(|key|, |keyID|)</dfn> method steps are:
+1. Let |promise| be [=a new promise=].
+1. If |keyID| is a {{bigint}} which cannot be represented as a integer between 0 and 2<sup>64</sup>-1 inclusive, [=reject=] |promise| with a {{RangeError}} exception, and abort these steps..
```
```suggestion
1. If |keyID| is a {{bigint}} which cannot be represented as a integer between 0 and 2<sup>64</sup>-1 inclusive, [=reject=] |promise| with a {{RangeError}} exception, and abort these steps.
```

youennf committed:
Update index.bs

youennf commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
 };
 
+interface mixin SFrameDecrypterManagement {
+    Promise<undefined> addDecryptionKey(CryptoKey key, CryptoKeyID keyID);
```
```suggestion
    Promise<undefined> addDecryptionKey(CryptoKey key, CryptoKeyID keyId);
```

youennf commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
 
+interface mixin SFrameDecrypterManagement {
+    Promise<undefined> addDecryptionKey(CryptoKey key, CryptoKeyID keyID);
+    Promise<undefined> removeDecryptionKey(CryptoKeyID keyID);
```
```suggestion
    Promise<undefined> removeDecryptionKey(CryptoKeyID keyId);
```

youennf commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
 
-interface mixin SFrameKeyManagement {
+interface mixin SFrameEncrypterManagement {
     Promise<undefined> setEncryptionKey(CryptoKey key, optional CryptoKeyID keyID);
```
```suggestion
    Promise<undefined> setEncryptionKey(CryptoKey key, optional CryptoKeyID keyId);
```

youennf committed:
Update index.bs

youennf committed:
Update index.bs

youennf committed:
Update index.bs

youennf commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
 
 ## Methods ## {#sframe-transform-methods}
-The <dfn method for="SFrameTransform">setEncryptionKey(|key|, |keyID|)</dfn> method steps are:
+The <dfn method for="SFrameEncrypterKeyManager">setEncryptionKey(|key|, |keyID|)</dfn> method steps are:
```
```suggestion
The <dfn method for="SFrameEncrypterKeyManager">setEncryptionKey(|key|, |keyId|)</dfn> method steps are:
```

youennf commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+   1. Let |currentKeyId| be |this|.`[[currentKeyId]]` if not undefined or 0 otherwise.
+   1. If |currentKeyId| is greater or equal to 2<sup>64</sup>-1, [=reject=] |promise| with a {{RangeError}} exception and abort these steps.
+   1. Set |keyId| to |currentKeyId| incremented by 1.
+1. If |keyID| is a {{bigint}} which cannot be represented as a integer between 0 and 2<sup>64</sup>-1 inclusive, [=reject=] |promise| with a {{RangeError}} exception and abort these steps.
```
```suggestion
1. If |keyId| is a {{bigint}} which cannot be represented as a integer between 0 and 2<sup>64</sup>-1 inclusive, [=reject=] |promise| with a {{RangeError}} exception and abort these steps.
```

youennf commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+1. If |keyID| is a {{bigint}} which cannot be represented as a integer between 0 and 2<sup>64</sup>-1 inclusive, [=reject=] |promise| with a {{RangeError}} exception and abort these steps.
+1.  Set |this|.`[[currentKeyId]]` to |keyId|.
+1. [=In parallel=], run the following steps:
+    1. Set the SFrame transform encryption algorithm's key material to |key| and |keyID|, as defined by [[RFC9605]].
```
```suggestion
    1. Set the SFrame transform encryption algorithm's key material to |key| and |keyId|, as defined by [[RFC9605]].
```

youennf commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+    1. [=Queue a task=] to [=resolve=] |promise| with undefined.
+1. Return |promise|.
+
+The <dfn method for="SFrameDecrypterKeyManager">addDecryptionKey(|key|, |keyID|)</dfn> method steps are:
```
```suggestion
The <dfn method for="SFrameDecrypterKeyManager">addDecryptionKey(|key|, |keyId|)</dfn> method steps are:
```

jan-ivar replied:
...here

youennf commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+
+The <dfn method for="SFrameDecrypterKeyManager">addDecryptionKey(|key|, |keyID|)</dfn> method steps are:
+1. Let |promise| be [=a new promise=].
+1. If |keyID| is a {{bigint}} which cannot be represented as a integer between 0 and 2<sup>64</sup>-1 inclusive, [=reject=] |promise| with a {{RangeError}} exception, and abort these steps.
```
```suggestion
1. If |keyId| is a {{bigint}} which cannot be represented as a integer between 0 and 2<sup>64</sup>-1 inclusive, [=reject=] |promise| with a {{RangeError}} exception, and abort these steps.
```

youennf commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+    1. [=Queue a task=] to [=resolve=] |promise| with undefined.
+1. Return |promise|.
+
+The <dfn method for="SFrameDecrypterKeyManager">removeDecryptionKey(|key|, |keyID|)</dfn> method steps are:
```
```suggestion
The <dfn method for="SFrameDecrypterKeyManager">removeDecryptionKey(|key|, |keyId|)</dfn> method steps are:
```

youennf commented on index.bs:
```diff
@@ -324,14 +343,38 @@ The <dfn>SFrame transform algorithm</dfn>, given |this| and |frame|, runs these
+
+The <dfn method for="SFrameDecrypterKeyManager">removeDecryptionKey(|key|, |keyID|)</dfn> method steps are:
+1. Let |promise| be [=a new promise=].
+1. If |keyID| is a {{bigint}} which cannot be represented as a integer between 0 and 2<sup>64</sup>-1 inclusive, [=reject=] |promise| with a {{RangeError}} exception, and abort these steps.
```
```suggestion
1. If |keyId| is a {{bigint}} which cannot be represented as a integer between 0 and 2<sup>64</sup>-1 inclusive, [=reject=] |promise| with a {{RangeError}} exception, and abort these steps.
```

jan-ivar commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
 
-interface mixin SFrameKeyManagement {
-    Promise<undefined> setEncryptionKey(CryptoKey key, optional CryptoKeyID keyID);
+interface mixin SFrameEncrypterManagement {
```
Pick SFrameEncrypterManagement or SFrameDecrypterKeyManager to match up with below...

youennf replied:
done

youennf commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
-interface mixin SFrameKeyManagement {
-    Promise<undefined> setEncryptionKey(CryptoKey key, optional CryptoKeyID keyID);
+interface mixin SFrameEncrypterManagement {
+    Promise<undefined> setEncryptionKey(CryptoKey key, optional CryptoKeyID keyId);
```
```suggestion
    Promise<undefined> setEncryptionKey(CryptoKey key, CryptoKeyID keyId);
```

youennf commented:
I'll update the PR to make the setEncryptionKey keyID parameter mandatory. This simplifies things and should be good enough.

youennf commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
 
-interface mixin SFrameKeyManagement {
-    Promise<undefined> setEncryptionKey(CryptoKey key, optional CryptoKeyID keyID);
+interface mixin SFrameEncrypterManagement {
```
```suggestion
interface mixin SFrameEncrypterManager {
```

youennf commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
     attribute EventHandler onerror;
 };
 
+interface mixin SFrameDecrypterManagement {
```
```suggestion
interface mixin SFrameDecrypterManager {
```

youennf committed:
Update index.bs

youennf committed:
Update index.bs

youennf commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
     constructor(optional SFrameTransformOptions options = {});
 };
-SFrameTransform includes SFrameKeyManagement;
+SFrameReceiverTransform includes SFrameDecrypterManagement;
```
```suggestion
SFrameReceiverTransform includes SFrameDecrypterManager;
```

youennf committed:
Update index.bs

youennf commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
+interface SFrameSenderTransform {
+    constructor(optional SFrameTransformOptions options = {});
+};
+SFrameSenderTransform includes SFrameEncrypterManagement;
```
```suggestion
SFrameSenderTransform includes SFrameEncrypterManager;
```

youennf committed:
Update index.bs

youennf commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
 };
 SFrameEncrypterStream includes GenericTransformStream;
-SFrameEncrypterStream includes SFrameKeyManagement;
+SFrameEncrypterStream includes SFrameEncrypterManagement;
```
```suggestion
SFrameEncrypterStream includes SFrameEncrypterManager;
```

youennf committed:
Update index.bs

youennf commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
 };
 SFrameDecrypterStream includes GenericTransformStream;
-SFrameDecrypterStream includes SFrameKeyManagement;
+SFrameDecrypterStream includes SFrameDecrypterManagement;
```
```suggestion
SFrameDecrypterStream includes SFrameDecrypterManager;
```

youennf committed:
Update index.bs

k-wasniowski commented on index.bs:
```diff
@@ -226,30 +227,42 @@ dictionary SFrameTransformOptions {
+};
+
+[Exposed=Window]
+interface SFrameSenderTransform {
```
There was a question in the past if SFrameTransform should have a 'RTC' prefix as other webrtc related stuff. I believe the answer not to add here a prefix had a good explanation that sframe is something bigger than just webrtc. Is this still relevant in current situation when sframe transform got splitted to sender/receiver? For me it in some way points now that webrtc will be main user of this transform.

youennf replied:
It makes sense to rename to RTCSFrameTransfrom, since it can only be used in RTC context now.
Let's do this in a follow-up though.

youennf replied:
Uploaded https://github.com/w3c/webrtc-encoded-transform/pull/291

youennf committed:
Update index.bs

youennf committed:
Update index.bs

youennf committed:
Update index.bs

youennf committed:
Update index.bs

youennf committed:
Update index.bs

youennf committed:
Update index.bs

youennf committed:
Update index.bs

youennf committed:
Update index.bs

youennf committed:
Fix links

youennf committed:
Remove keyId undefined handling

youennf committed:
Use infra remove like done for set
